rust/actyx/ax-aql/src/language/types.rs

```diff
@@ -25,6 +33,319 @@ pub enum Type {
+            (Type::Atom(TypeAtom::Universal), _) => Hierarchy::Supertype,
+            (_, Type::Atom(TypeAtom::Universal)) => Hierarchy::Subtype,
+            (Type::Atom(a), Type::Atom(b)) => match (a, b) {
+                // supertypes
+                (TypeAtom::Bool(None), TypeAtom::Bool(Some(_))) => Hierarchy::Supertype,
+                (TypeAtom::Number(None), TypeAtom::Number(Some(_))) => Hierarchy::Supertype,
+                (TypeAtom::String(None), TypeAtom::String(Some(_))) => Hierarchy::Supertype,
+                // subtypes
+                (TypeAtom::Bool(Some(_)), TypeAtom::Bool(None)) => Hierarchy::Subtype,
+                (TypeAtom::Number(Some(_)), TypeAtom::Number(None)) => Hierarchy::Subtype,
+                (TypeAtom::String(Some(_)), TypeAtom::String(None)) => Hierarchy::Subtype,
+                _ => Hierarchy::Disjointed,
+            },
+            (Type::Record(a), Type::Record(b)) => {
+                let a = a.iter().collect::<BTreeSet<_>>();
+                let b = b.iter().collect::<BTreeSet<_>>();
+                let intersection = BTreeSet::intersection(&a, &b).cloned().collect::<BTreeSet<_>>();
+
+                if intersection == b {
+                    Hierarchy::Supertype
+                } else if intersection == a {
+                    Hierarchy::Subtype
+                } else {
+                    Hierarchy::Disjointed
+                }
+            }
+            (Type::Union(a), Type::Union(b)) => {
+                compare_unions(&Type::spread_union(a.as_ref()), &Type::spread_union(b.as_ref()))
+            }
+            (Type::Union(a), b) => compare_unions(&Type::spread_union(a.as_ref()), &BTreeSet::from([b])),
+            (a, Type::Union(b)) => compare_unions(&BTreeSet::from([a]), &Type::spread_union(b.as_ref())),
+            (Type::Array(a), Type::Array(b)) => a.hierarchy_towards(b),
+            (Type::Dict(a), Type::Dict(b)) => a.hierarchy_towards(b),
+            _ => Hierarchy::Disjointed,
+        }
+    }
+
+    /// Spread union tree into a set of type references without collapsing
+    fn spread_union((a, b): &(Type, Type)) -> BTreeSet<&Type> {
+        let spread = spread(vec![a, b], |x| match x {
+            Type::Union(union) => {
+                let (a, b) = union.as_ref();
+                Spread::Many(vec![a, b])
+            }
+            x => Spread::One(x),
+        });
+
+        spread.into_iter().collect()
+    }
+
+    /// Spread union tree into a set of types with collapsing
+    fn spread_union_collapsing((a, b): &(Type, Type)) -> BTreeSet<Type> {
+        let spread = spread(vec![a.clone(), b.clone()], |x| match x.clone().collapse() {
+            Type::Union(union) => {
+                let (a, b) = union.as_ref();
+                Spread::Many(vec![a.clone(), b.clone()])
+            }
+            x => Spread::One(x),
+        });
+
+        spread.into_iter().collect()
+    }
+}
```

Comment:
I feel there must be a better way of writing this API. The double clone that ends up happening in line 337 seems wrong.

Maybe you could use a `Collapse` trait and implement it for the various `BTreeSet`, delaying the cloning until its really necessary for example. Furthermore, as we discussed, `collapse` probably doesn't need to consume the type, you're just checking a bunch of things and returning a new type.

Reply:
while collapse doesn't consume, it still produces a new type.
this in turn forces this function to return BTreeSet<Type> and not BTreeSet<&Type>
because we cannot return references to objects created within the function (produced by collapse)

Reply:
Yeah that makes sense but my point was also about how (if possible) to make a slightly better API by just having a single function instead of these two.